src/main/flight/imu.c

```diff
@@ -201,86 +201,101 @@ static float imuUseFastGains(void) {
+    quaternionNormalize(vAcc);
+    // Error is sum of cross product between estimated direction and measured direction of gravity
+    vError->x += (vAcc->y * (1.0f - 2.0f * qpAttitude.xx - 2.0f * qpAttitude.yy) - vAcc->z * (2.0f * (qpAttitude.yz - -qpAttitude.wx)));
+    vError->y += (vAcc->z * (2.0f * (qpAttitude.xz + -qpAttitude.wy)) - vAcc->x * (1.0f - 2.0f * qpAttitude.xx - 2.0f * qpAttitude.yy));
+    vError->z += (vAcc->x * (2.0f * (qpAttitude.yz - -qpAttitude.wx)) - vAcc->y * (2.0f * (qpAttitude.xz + -qpAttitude.wy)));
+}
 
-    if (useCOG) {
-        // Use raw heading error (from GPS or whatever else)
-        while (courseOverGround >  M_PIf) courseOverGround -= (2.0f * M_PIf);
-        while (courseOverGround < -M_PIf) courseOverGround += (2.0f * M_PIf);
+FAST_CODE float calcMagError(quaternion *vMag){
+    quaternionNormalize(vMag);
+    
+    // (hx; hy; 0) - measured mag field vector in EF (assuming Z-component is zero)
+    // (bx; 0; 0) - reference mag field vector heading due North in EF (assuming Z-component is zero)
+    const float hx = (1.0f - 2.0f * qpAttitude.yy - 2.0f * qpAttitude.zz) * vMag->x + (2.0f * (qpAttitude.xy + -qpAttitude.wz)) * vMag->y + (2.0f * (qpAttitude.xz - -qpAttitude.wy)) * vMag->z;
+    const float hy = (2.0f * (qpAttitude.xy - -qpAttitude.wz)) * vMag->x + (1.0f - 2.0f * qpAttitude.xx - 2.0f * qpAttitude.zz) * vMag->y + (2.0f * (qpAttitude.yz + -qpAttitude.wx)) * vMag->z;
+    const float bx = sqrtf(sq(hx) + sq(hy));
+
+    // magnetometer error is cross product between estimated magnetic north and measured magnetic north (calculated in EF)
+    return -(float)(hy * bx);
+}
+FAST_CODE float calcCogError(float courseOverGround){
+    // Use raw heading error (from GPS or whatever else)
+    while (courseOverGround >  M_PIf) courseOverGround -= (2.0f * M_PIf);
+    while (courseOverGround < -M_PIf) courseOverGround += (2.0f * M_PIf);
+
+    // William Premerlani and Paul Bizard, Direction Cosine Matrix IMU - Eqn. 22-23
+    // (Rxx; Ryx) - measured (estimated) heading vector (EF)
+    // (cos(COG), sin(COG)) - reference heading vector (EF)
+    // error is cross product between reference heading and estimated heading (calculated in EF)
+    return -(float)sin_approx(courseOverGround) * (1.0f - 2.0f * qpAttitude.yy - 2.0f * qpAttitude.zz) - cos_approx(courseOverGround) * (2.0f * (qpAttitude.xy - -qpAttitude.wz));   
+}
 
-        // William Premerlani and Paul Bizard, Direction Cosine Matrix IMU - Eqn. 22-23
-        // (Rxx; Ryx) - measured (estimated) heading vector (EF)
-        // (cos(COG), sin(COG)) - reference heading vector (EF)
-        // error is cross product between reference heading and estimated heading (calculated in EF)
-        const float ez_ef = - sin_approx(courseOverGround) * (1.0f - 2.0f * qpAttitude.yy - 2.0f * qpAttitude.zz) - cos_approx(courseOverGround) * (2.0f * (qpAttitude.xy - -qpAttitude.wz));
+FAST_CODE void applySensorCorrection(quaternion *vError){
+    UNUSED(vError);
```

Comment:
Implicit declaration is not allowed and is causing Trav to throw error on the TEST build